File: valid-parentheses/valid.c

#include <stdio.h>
#include <string.h>

static int
valid_parens(char *buf)
{
  static char stack[1024];

  char *p = buf;
  int size = 0;
  for (; *p; ++p) {
    switch (*p) {
      case '(': case '{': case '[':
        stack[size++] = *p;
        break;
      case ')':
        if (stack[--size] != '(')
          return 0;
        break;
      case ']':
        if (stack[--size] != '[')
          return 0;
        break;
      case '}':
        if (stack[--size] != '{')
          return 0;
        break;
    }
  }
  return size == 0;
}

int
main(int argc, char const *argv[])
{
  FILE *file = fopen(argv[1], "r");
  char line[1024];
  while (fgets(line, 1024, file)) {
    if (strlen(line) > 0)
      printf("%s\n", valid_parens(line) ? "True" : "False");
  }
  return 0;
}

File: valid-parentheses/valid.rb
module ValidParentheses
  class ParenStack
    PARENS = {')' => '(', ']' => '[', '}' => '{'}

    def initialize
      @valid = true
      @stack = []
    end

    def <<(item)
      return unless @valid

      if PARENS.key? item
        if @stack[-1] == PARENS[item]
          @stack.pop
        else
          @valid = false
        end
      else
        @stack << item
      end
    end

    def valid?
      @valid && @stack.length == 0
    end
  end

  def self.valid_parentheses?(line)
    stack = ParenStack.new
    line.each_char do |char|
      stack << char
    end
    stack.valid?
  end

  def self.run(filename = ARGV[0])
    File.open(filename) do |file|
      file.each_line do |line|
        line.strip!
        next if line.empty?

        puts valid_parentheses?(line) ? 'True' : 'False'
      end
    end
  end
end

ValidParentheses.run if __FILE__ == $0